### PR TITLE
Parser cleanup

### PR DESCRIPTION
**Pull request by lushoBarlett**
Attempts to reduce many lines from the parser, without a change in speed. Basically a tiny refactor.

Maybe the littlest speed might get sacrificed in order to abstract some bits into auxiliary functions.

### CODE REVIEW

**Comment by lushoBarlett** [comment]
Further improvements will be added, mainly while loops and some common parse patterns. I want to merge #200 first because there's a new function there that I wanted to use here.

**Comment by SebastianMestre** [comment]
Hey. I would recommend copying the function from the other branch, and dealing with possible conflicts at merge time.

**Comment by lushoBarlett** [comment]
Fair point. I still have to decide what other changes I can do. Any ideas you have are welcome

File: src/parser.cpp [hunk 1]
```diff
@@ -8,6 +8,14 @@
 
 #include <cassert>
 
+#define HANDLE(writer)                                                         \
+	if (handle_error(result, writer))                                      \
```
Comment (SebastianMestre) [inline src/parser.cpp:12]: i dont like the name `HANDLE`, maybe something more descriptive, like `HANDLE_ERROR` or `CHECK_AND_RETURN`, like we use elsewhere, would be better.

File: src/parser.cpp [hunk 2]
```diff
@@ -8,6 +8,14 @@
...
+		return result;
+
+#define REQUIRE(token)                                                         \
+	if (handle_error(result, require(TokenTag::token)))                    \
+		return result;
```
Comment (SebastianMestre) [inline src/parser.cpp:17]: I think I prefer typing `TokenTag::` at the usage site. wdyt?

Comment (SebastianMestre) [reply src/parser.cpp:17]: It also allows for things like

```c++
auto tag = TokenTag::something;
REQUIRE(tag);
```

**Comment by SebastianMestre** [comment]
I'm not too sure about 6480327

**Comment by lushoBarlett** [comment]
I was very unsure of some of my decisions here, so I will gladly accept your suggestions and apply them sometime soon

**Comment by lushoBarlett** [comment]
I reverted the bad parts, and I'm gonna stop right here, because I have no clue what else to do. If it needs more cleanup, another pr can be done.

**Review by SebastianMestre — APPROVED** [review]
This looks much better to me

**Pull request merged**